ext/src/helpers/output_limited_buffer.rs

```diff
@@ -0,0 +1,50 @@
+use magnus::{value::InnerValue, value::Opaque, RString, Ruby};
+use std::io;
+
+pub struct OutputLimitedBuffer {
+    buffer: Opaque<RString>,
+    /// The maximum number of bytes that can be written to the output stream buffer.
+    capacity: usize,
+}
+
+impl OutputLimitedBuffer {
+    #[must_use]
+    pub fn new(buffer: Opaque<RString>, capacity: usize) -> Self {
```

Comment:
> have you considered inheriting stdout / stderr and moving the contents to a String?

Do you mean when running the wasm in ruby?

Reply:
Something like:

```ruby
def capture_stdout
  original_stdout = $stdout  
  $stdout = StringIO.new    
  yield                    
  $stdout.string          
ensure
  $stdout = original_stdout 
end
```

and using it as:

```ruby
res = capture_stdout {
  instance = linker.instantiate(store, mod)
  instance.invoke("_start")
}

puts res
```

One thing to note is that I haven't deeply thought about all the implications of doing this though, it's totally possible that this is not a great approach for your use-case .

---

Thinking a bit more about my previous comment and thinking about concrete next steps here:

I've confirmed that Magnus [should be updated](https://docs.rs/magnus/latest/src/magnus/r_string.rs.html#1630-1634) to gracefully handle errors when writing to a `RString` (maybe we also need to check if this should be updated for all other types that: (i) implement `Write` and (ii) can be frozen).

I've also confirmed on top of your branch that freezing a String *after* calling `set_{stdout,stderr}_buffer` should be fine as along as we check if the object is frozen before writing into it:

```diff
diff --git a/ext/src/helpers/output_limited_buffer.rs b/ext/src/helpers/output_limited_buffer.rs
index 50c98c1..845442c 100644
--- a/ext/src/helpers/output_limited_buffer.rs
+++ b/ext/src/helpers/output_limited_buffer.rs
@@ -1,4 +1,8 @@
-use magnus::{value::InnerValue, value::Opaque, RString, Ruby};
+use magnus::{
+    value::Opaque,
+    value::{InnerValue, ReprValue},
+    RString, Ruby,
+};
 use std::io;
 
 pub struct OutputLimitedBuffer {
@@ -22,6 +26,13 @@ impl io::Write for OutputLimitedBuffer {
 
         let mut inner_buffer = self.buffer.get_inner_with(&ruby);
 
+        let is_frozen = inner_buffer.as_value().is_frozen();
+
+        if is_frozen {
+            dbg!("is frozen");
+            return Ok(0);
+        }
+
         if buf.is_empty() {
             return Ok(buf.len());
         }
diff --git a/spec/unit/wasi_spec.rb b/spec/unit/wasi_spec.rb
index fe30d50..e8e7a69 100644
--- a/spec/unit/wasi_spec.rb
+++ b/spec/unit/wasi_spec.rb
@@ -66,11 +66,12 @@ module Wasmtime
           .set_stderr_buffer(stderr_str, 40000)
           .build
 
+        stdout_str.freeze
         run_wasi_module(wasi_config)
 
-        parsed_stdout = JSON.parse(stdout_str)
+        parsed_stdout = ''
         parsed_stderr = JSON.parse(stderr_str)
-        expect(parsed_stdout.fetch("name")).to eq("stdout")
+        expect("").to eq("")
         expect(parsed_stderr.fetch("name")).to eq("stderr")
       end
 
``` 
The ideal implementation for this feature IMHO, would be to:

1. Make a PR to Magnus to ensure that any `write` errors are gracefully handled.
2. With graceful error handling out of the way, find a way to correctly store any errors that may occur outside of Wasm execution[^1].
3. Use `StringIO` instead of `String` as the output buffer for both `std_{out,err}`.

That said though, you're not expected to handle everything above, having verified that: the frozen case can be handled and that Ruby doesn't panic when invalid encoding is written to the String, I'd say it's fine to land this PR with the following changes:

1. Check if the String is frozen before writing into it and return `0` bytes written if the String is frozen.
2. Document the implications of freezing strings that are used as stdout/stderr buffers.
3. Document that no encoding checks are done in the resulting string, therefore it means that it's the callers responsibility to ensure that the String contains a valid encoding if necessary.
4. Add tests around the usage of frozen strings. 

I'll create an issue to track the follow-up work needed (use `StringIO`, make the necessary fixes in magnus, etc). 

[^1]: We need to store the error that gets returned from Magnus somewhere in order to report it later on. Currently, we store the last error in Wasmtime's store: https://github.com/bytecodealliance/wasmtime-rb/blob/main/ext/src/ruby_api/store.rs#L53; in the context of this change, that is not  possible since we don't have a handle to the store, however, we could make use of a global store (like a thread local or something like that).

Reply:
Thanks for the help, I'll make those changes

Reply:
I made the change to return 0 when the string is frozen, only it seems to work for stdout and not stderr. For stderr it looks like it retries a few times and then throws a wasm trap. Do you know if there is something special happening only for stderr where it keeps retrying?